torchrl/envs/vec_env.py

```diff
@@ -1076,3 +1092,297 @@ def _run_worker_pipe_shared_mem(
             else:
                 # don't send env through pipe
                 child_pipe.send(("_".join([cmd, "done"]), None))
+
+
+class MultiThreadedEnvWrapper(_EnvWrapper):
+    """Wrapper for envpool-based multithreaded environments."""
+
+    _verbose: bool = False
+
+    def __init__(
+        self,
+        env: Optional["envpool.python.envpool.EnvPoolMixin"] = None,
+        **kwargs,
+    ):
+        if not _has_envpool:
+            raise ImportError(
+                f"""envpool python package or one of its dependencies (gym, treevalue) were not found. Please install these dependencies.
+(Got the error message: {IMPORT_ERR_ENVPOOL}).
+"""
+            )
+        if env is not None:
+            kwargs["env"] = env
+            self.num_workers = env.config["num_envs"]
+            # For synchronous mode batch size is equal to the number of workers
+            self.batch_size = torch.Size([self.num_workers])
+        super().__init__(**kwargs)
+
+        # Buffer to keep the latest observation for each worker
+        # It's a TensorDict when the observation consists of several variables, e.g. "position" and "velocity"
+        self.obs: Union[torch.tensor, TensorDict] = self.observation_spec[
+            "observation"
+        ].zero((self.num_workers,))
+
+    def _check_kwargs(self, kwargs: Dict):
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, (envpool.python.envpool.EnvPoolMixin,)):
+            raise TypeError("env is not of type 'envpool.python.envpool.EnvPoolMixin'.")
+
+    def _build_env(self, env: "envpool.python.envpool.EnvPoolMixin"):
+        return env
+
+    def _make_specs(
+        self, env: "envpool.python.envpool.EnvPoolMixin"
+    ) -> None:  # noqa: F821
+        self.input_spec = self._get_input_spec()
+        self.reward_spec = self._get_reward_spec()
+        self.observation_spec = self._get_observation_spec()
+
+    def _init_env(self) -> Optional[int]:
+        pass
+
+    def _reset(self, tensordict: TensorDictBase) -> TensorDictBase:
+        reset_workers = self._parse_reset_workers(tensordict)
+        reset_data = self._env.reset(reset_workers)
+        tensordict_out = self._transform_reset_output(reset_data, reset_workers)
+        self.is_closed = False
+        return tensordict_out
+
+    def _step(self, tensordict: TensorDictBase) -> TensorDictBase:
+        action = tensordict.get("action")
+        # Action needs to be moved to CPU and converted to numpy before being passed to envpool
+        action = action.to(torch.device("cpu"))
+        step_output = self._env.step(action.detach().numpy())
```

Comment:
I think this `detach` is redundant since the tensor is already on CPU. Not a big deal though

Reply:
Removed `.detach()`, added the decorator. One of them is needed, otherwise it throws "_RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead_" even on CPU